### DISCUSSION TITLE
Eclipse, Metals, Nexus with authentications

### DISCUSSION THREAD

**Discussion by MarkusLupusWolf** [scalameta/metals#2442] [General]
Hi,

I am trying to get Metals to work with Eclipse and our Setup.
First, we are behind a firewall and need proxy connections to the internet.
I tried to configure coursier to use our Nexus repository which mirrors maven-central. This essentially worked, but our Nexus uses Authorization and I only get the following message when Metals tries to start the language server:

error: The command bsp --protocol tcp --port 41321 returned with an error
> No server running at 127.0.0.1:8212, let's fire one...
> Resolving ch.epfl.scala:bloop-frontend_2.12:1.4.6-21-464e4ec4...
> error: Unexpected error when resolving Bloop server via coursier!
> error: Error downloading ch.epfl.scala:bloop-frontend_2.12:1.4.6-21-464e4ec4
>   not found: C:\Users\A303609\.ivy2\local\ch.epfl.scala\bloop-frontend_2.12\1.4.6-21-464e4ec4\ivys\ivy.xml
>   unauthorized: https://artifactory.devops.telekom.de/artifactory/maven-central/ch/epfl/scala/bloop-frontend_2.12/1.4.6-21-464e4ec4/bloop-frontend_2.12-1.4.6-21-464e4ec4.pom (Artifactory Realm)

I looked at the coursier documentation and tried it with a credentials.properties file containing username and password, but it seems not to use this file. I put the file in the same location as the mirror.properties file.

Any hints, helps, documentation on debugging, etc. would be appreciated.

Thanks,

Markus

**Comment by tgodzik** [comment]
I am trying to update couriser inside Bloop, which might help in this situation, but for now could you maybe try to add `coursier.credentials` java property (with the file path) inside Bloop config directory? You can find more information here https://scalacenter.github.io/bloop/docs/server-reference#custom-java-options

**Reply by MarkusLupusWolf** [reply]
Thanks very much Tomasz, that was a good hint, but it did not work.
However, it could be that I am doing something wrong, being on a Windows machine:.

I created a .bloop dir in C:\Users\<myUsername>. 
In this directory I created a file bloop.json with the following content:

{
  "javaOptions": ["-Dcoursier.credentials=C:\\Users\\<myUserName>\\AppData\\Roaming\\Coursier\\config\\coursier.credentials"]
}

I tried with \\ and with \ in the path just to be sure. Both did not work.  It could also be that the location of the .bloop directory is wrong. Is there a possibility to get more debugging information on what is happening in the plugin?

**Reply by tgodzik** [reply]
It should be possible to see in the task manager whether the Bloop process is picking up the right jvm options. It might also be an issue of an outdated coursier versions used in Bloop. You should be able to change the Bloop version in the Metals settings to `1.4.6-33-1c6f6712`

I think the place to put the Bloop directory should be something like `${user.home}/.bloop/bloop.json` - it's a bit tricky to figure it out.

I was thinking of trying to have that all setup by Metals, but I have very little experience with how proxies should work and what are the possible different scenarios. If we get the full picture there, we could try to set it all up automatically.

**Reply by MarkusLupusWolf** [reply]
Thanks again, but I do not see any bloop process in the task manager. What I see in the Metals log is the following:
Starting the bsp launcher for bloop...
Opening a bsp server connection with 'bsp --protocol tcp --port 41321'...
Waiting for the bsp connection to come up...
...
Waiting for the bsp connection to come up...
error: The command bsp --protocol tcp --port 41321 returned with an error
> No server running at 127.0.0.1:8212, let's fire one...
> Resolving ch.epfl.scala:bloop-frontend_2.12:1.4.6-21-464e4ec4...
> error: Unexpected error when resolving Bloop server via coursier!
> error: Error downloading ch.epfl.scala:bloop-frontend_2.12:1.4.6-21-464e4ec4
>   not found: C:\Users\A303609\.ivy2\local\ch.epfl.scala\bloop-frontend_2.12\1.4.6-21-464e4ec4\ivys\ivy.xml
>   unauthorized: https://artifactory.devops.telekom.de/artifactory/maven-central/ch/epfl/scala/bloop-frontend_2.12/1.4.6-21-464e4ec4/bloop-frontend_2.12-1.4.6-21-464e4ec4.pom (Artifactory Realm)
>   download error: Caught java.net.UnknownHostException: dl.bintray.com (dl.bintray.com) while downloading https://dl.bintray.com/scalacenter/releases/ch/epfl/scala/bloop-frontend_2.12/1.4.6-21-464e4ec4/bloop-frontend_2.12-1.4.6-21-464e4ec4.pom
> bloop.shaded.coursier.error.ResolutionError$CantDownloadModule: Error downloading ch.epfl.scala:bloop-frontend_2.12:1.4.6-21-464e4ec4
>   not found: C:\Users\A303609\.ivy2\local\ch.epfl.scala\bloop-frontend_2.12\1.4.6-21-464e4ec4\ivys\ivy.xml
>   unauthorized: https://artifactory.devops.telekom.de/artifactory/maven-central/ch/epfl/scala/bloop-frontend_2.12/1.4.6-21-464e4ec4/bloop-frontend_2.12-1.4.6-21-464e4ec4.pom (Artifactory Realm)
>   download error: Caught java.net.UnknownHostException: dl.bintray.com (dl.bintray.com) while downloading https://dl.bintray.com/scalacenter/releases/ch/epfl/scala/bloop-frontend_2.12/1.4.6-21-464e4ec4/bloop-frontend_2.12-1.4.6-21-464e4ec4.pom
> 	at bloop.shaded.coursier.Resolve$.$anonfun$validate$1(Resolve.scala:325)

It could be that whatever is fired up does not live long enough to see what is happening. Can you explain what is being started here?

I checked the path with a small Java program. On Windows it points to C:\Users\<myUsername>, so that bloop.json file is at the right place.

**Reply by tgodzik** [reply]
I wonder if maybe it would help if you added the credentials to `.jvmopts` file in the workspace. You could also try the server properties setting, but I am not sure how to set that properly in Eclipse. Did you try to update the Bloop version? Thought that also need to go into settings.

The Eclipse plugin hasn't been updated in a while, so it might all be far from ideal.

**Comment by svenbraun** [comment]
I am experience the same issue with VSCode, Metals and Artifactory.

Does somebody know of a solution?

**Reply by tgodzik** [reply]
Did you try to look at https://scalameta.org/metals/docs/troubleshooting/proxy ?

**Reply by svenbraun** [reply]
Hello,

thanks for pointing to this page, that had helped me a lot before already ;-).

Unfortunately it does not describe the solution for my issue, which I experience in VSCode / metal too:

```
C:\Users\bsv2fr\Documents\Projekte\2020 TGSE\hello-world-20211122-1221>bloop server
Attempting a connection to the server...
Resolving ch.epfl.scala:bloop-frontend_2.12:1.4.11...
error: Unexpected error when resolving Bloop server via coursier!
error: Error downloading ch.epfl.scala:bloop-config_2.12:1.4.11
  not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\bloop-config_2.12\1.4.11\ivys\ivy.xml
  unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/bloop-config_2.12/1.4.11/bloop-config_2.12-1.4.11.pom (Artifactory Realm)
```
...similar download trials deleted
```
at coursier.error.ResolutionError$.from(ResolutionError.scala:117)
        at coursier.Resolve.validate0$1(Resolve.scala:117)
        at coursier.Resolve.$anonfun$ioWithConflicts0$4(Resolve.scala:161)
        at coursier.util.Task$.$anonfun$flatMap$extension$1(Task.scala:14)
        at coursier.util.Task$.$anonfun$flatMap$extension$1$adapted(Task.scala:14)
        at coursier.util.Task$.wrap(Task.scala:84)
        at coursier.util.Task$.$anonfun$flatMap$2(Task.scala:14)
        at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:307)
        at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:41)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
        at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:517)
        at com.oracle.svm.core.windows.WindowsJavaThreads.osThreadStartRoutine(WindowsJavaThreads.java:138)
error: Failed to locate server, aborting start of server!
```


To give you a little more **context**,  I have
* Credentials configured in C:\Users\bsv2fr\AppData\Roaming\Coursier\config\credentials.properties
* Mirror configured in C:\Users\bsv2fr\AppData\Roaming\Coursier\config\mirror.properties
* have cntlm running as local proxy
* configured proxy in
** environment variables
** .jvmopts
* metals.server Properties set for https.proxyHost, https.proxyPort, coursier.mirror, coursier.credentials
* metals.bloopsbtalreadyInstalled is off

coursier works, (only) when passing the proxy using the command line. It does _not_ work when -J parameters are _not_ provided.

```
C:\Users\bsv2fr\Documents\Projekte\2020 TGSE\hello-world-20211122-1221>cs -J-Dhttps.proxyHost=127.0.0.1 -J-Dhttps.proxyPort=3128 install bloop
https://rb-artifactory.bosch.com:443/artifactory/remote-repos/io/get-coursier/apps/maven-metadata.xml
  No new update since 2021-11-16 01:21:19
https://repo1.maven.org/maven2/org/scala-lang/scala-library/maven-metadata.xml
  No new update since 2021-11-01 16:11:58
https://repo1.maven.org/maven2/ch/epfl/scala/bloopgun_2.12/maven-metadata.xml
  No new update since 2021-11-02 20:11:03
https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/bloopgun_2.12/maven-metadata.xml
  No new update since 2021-11-10 21:48:08

C:\Users\bsv2fr\Documents\Projekte\2020 TGSE\hello-world-20211122-1221>cs -J-Dhttps.proxyHost=127.0.0.1 -J-Dhttps.proxyPort=3128 setup
Checking if a JVM is installed
Found a JVM installed under C:\Program Files\Java\jdk1.8.0_311.


Checking if ~\AppData\Local\Coursier\data\bin is in PATH

Checking if the standard Scala applications are installed
  Found ammonite
  Found cs
  Found coursier
  Found scala
  Found scalac
  Found sbt
  Found sbtn
  Found scalafmt
```

Here is the excerpt of metals.log:
```
[0m2021.11.22 18:54:52 INFO  Attempting to connect to the build server...[0m
Starting the bsp launcher for bloop...
Opening a bsp server connection with 'bsp --protocol tcp --port 18073'...
Waiting for the bsp connection to come up...
Waiting for the bsp connection to come up...
Waiting for the bsp connection to come up...
error: The command bsp --protocol tcp --port 18073 returned with an error
> No server running at 127.0.0.1:8212, let's fire one...
> Resolving ch.epfl.scala:bloop-frontend_2.12:1.4.11...
> error: Unexpected error when resolving Bloop server via coursier!
> error: Error downloading ch.epfl.scala:bloop-config_2.12:1.4.11
>   not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\bloop-config_2.12\1.4.11\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/bloop-config_2.12/1.4.11/bloop-config_2.12-1.4.11.pom (Artifactory Realm)
> Error downloading ch.epfl.scala:bloop-shared_2.12:1.4.11
>   not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\bloop-shared_2.12\1.4.11\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/bloop-shared_2.12/1.4.11/bloop-shared_2.12-1.4.11.pom (Artifactory Realm)
> Error downloading org.scalaz:scalaz-core_2.12:7.2.20
>   not found: C:\Users\bsv2fr\.ivy2\local\org.scalaz\scalaz-core_2.12\7.2.20\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/org/scalaz/scalaz-core_2.12/7.2.20/scalaz-core_2.12-7.2.20.pom (Artifactory Realm)
> Error downloading ch.epfl.scala:bloop-backend_2.12:1.4.11
>   not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\bloop-backend_2.12\1.4.11\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/bloop-backend_2.12/1.4.11/bloop-backend_2.12-1.4.11.pom (Artifactory Realm)
> Error downloading io.monix:monix_2.12:2.3.3
>   not found: C:\Users\bsv2fr\.ivy2\local\io.monix\monix_2.12\2.3.3\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/io/monix/monix_2.12/2.3.3/monix_2.12-2.3.3.pom (Artifactory Realm)
> Error downloading com.github.alexarchambault:case-app_2.12:2.0.6
>   not found: C:\Users\bsv2fr\.ivy2\local\com.github.alexarchambault\case-app_2.12\2.0.6\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/com/github/alexarchambault/case-app_2.12/2.0.6/case-app_2.12-2.0.6.pom (Artifactory Realm)
> Error downloading ch.epfl.scala:sockets:1.4.11
>   not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\sockets\1.4.11\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/sockets/1.4.11/sockets-1.4.11.pom (Artifactory Realm)
> Error downloading ch.epfl.scala:scala-debug-adapter_2.12:2.0.11
>   not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\scala-debug-adapter_2.12\2.0.11\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/scala-debug-adapter_2.12/2.0.11/scala-debug-adapter_2.12-2.0.11.pom (Artifactory Realm)
> bloop.shaded.coursier.error.ResolutionError$Several: Error downloading ch.epfl.scala:bloop-config_2.12:1.4.11
>   not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\bloop-config_2.12\1.4.11\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/bloop-config_2.12/1.4.11/bloop-config_2.12-1.4.11.pom (Artifactory Realm)
> Error downloading ch.epfl.scala:bloop-shared_2.12:1.4.11
>   not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\bloop-shared_2.12\1.4.11\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/bloop-shared_2.12/1.4.11/bloop-shared_2.12-1.4.11.pom (Artifactory Realm)
> Error downloading org.scalaz:scalaz-core_2.12:7.2.20
>   not found: C:\Users\bsv2fr\.ivy2\local\org.scalaz\scalaz-core_2.12\7.2.20\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/org/scalaz/scalaz-core_2.12/7.2.20/scalaz-core_2.12-7.2.20.pom (Artifactory Realm)
> Error downloading ch.epfl.scala:bloop-backend_2.12:1.4.11
>   not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\bloop-backend_2.12\1.4.11\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/bloop-backend_2.12/1.4.11/bloop-backend_2.12-1.4.11.pom (Artifactory Realm)
> Error downloading io.monix:monix_2.12:2.3.3
>   not found: C:\Users\bsv2fr\.ivy2\local\io.monix\monix_2.12\2.3.3\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/io/monix/monix_2.12/2.3.3/monix_2.12-2.3.3.pom (Artifactory Realm)
> Error downloading com.github.alexarchambault:case-app_2.12:2.0.6
>   not found: C:\Users\bsv2fr\.ivy2\local\com.github.alexarchambault\case-app_2.12\2.0.6\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/com/github/alexarchambault/case-app_2.12/2.0.6/case-app_2.12-2.0.6.pom (Artifactory Realm)
> Error downloading ch.epfl.scala:sockets:1.4.11
>   not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\sockets\1.4.11\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/sockets/1.4.11/sockets-1.4.11.pom (Artifactory Realm)
> Error downloading ch.epfl.scala:scala-debug-adapter_2.12:2.0.11
>   not found: C:\Users\bsv2fr\.ivy2\local\ch.epfl.scala\scala-debug-adapter_2.12\2.0.11\ivys\ivy.xml
>   unauthorized: https://rb-artifactory.bosch.com:443/artifactory/remote-repos/ch/epfl/scala/scala-debug-adapter_2.12/2.0.11/scala-debug-adapter_2.12-2.0.11.pom (Artifactory Realm)
> 	at bloop.shaded.coursier.error.ResolutionError$.from(ResolutionError.scala:117)
> 	at bloop.shaded.coursier.Resolve.validate0$1(Resolve.scala:117)
> 	at bloop.shaded.coursier.Resolve.$anonfun$ioWithConflicts0$4(Resolve.scala:161)
> 	at bloop.shaded.coursier.util.Task$.$anonfun$flatMap$extension$1(Task.scala:14)
> 	at bloop.shaded.coursier.util.Task$.$anonfun$flatMap$extension$1$adapted(Task.scala:14)
> 	at bloop.shaded.coursier.util.Task$.wrap(Task.scala:84)
> 	at bloop.shaded.coursier.util.Task$.$anonfun$flatMap$2(Task.scala:14)
> 	at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:307)
> 	at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:41)
> 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
> 	at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
> 	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
> 	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1067)
> 	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1703)
> 	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:172)
> error: Server could not be started at None:Some(8212)! Giving up...
error: The launcher failed to establish a bsp connection, aborting...
[0m2021.11.22 18:54:56 ERROR Failed to connect with build server, no functionality will work.
java.lang.RuntimeException: The server did not start, got FailedToOpenBspConnection
	at bloop.launcher.LauncherMain.failPromise$1(Launcher.scala:95)
	at bloop.launcher.LauncherMain.runLauncher(Launcher.scala:122)
	at scala.meta.internal.metals.BloopServers$$anon$1.run(BloopServers.scala:159)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[0m
```

**Reply by tgodzik** [reply]
Thanks for posting all the details, it could help when working on https://github.com/scalameta/metals/issues/2485

Could you try adding `-J-Dhttps.proxyHost=127.0.0.1 -J-Dhttps.proxyPort=3128` to https://scalacenter.github.io/bloop/docs/server-reference#custom-java-options

I think on windows it should be the user directory then `.bloop/bloop.json`

I think would be awesome add add an option in VS Code so that everything is configured for you. We should take a look at it
